x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/FrozenStorageDeciderService.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.autoscaling.storage;
+
+import org.elasticsearch.cluster.ClusterInfo;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.node.DiscoveryNodeRole;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingCapacity;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderContext;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderResult;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderService;
+import org.elasticsearch.xpack.autoscaling.util.FrozenUtils;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.IntStream;
+import java.util.stream.StreamSupport;
+
+public class FrozenStorageDeciderService implements AutoscalingDeciderService {
+    public static final String NAME = "frozen_storage";
+
+    static final double DEFAULT_PERCENTAGE = 5.0d;
+    public static final Setting<Double> PERCENTAGE = Setting.doubleSetting("percentage", DEFAULT_PERCENTAGE, 0.0d);
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        Metadata metadata = context.state().metadata();
+        long dataSetSize = StreamSupport.stream(metadata.spliterator(), false)
+            .filter(imd -> FrozenUtils.isFrozenIndex(imd.getSettings()))
+            .mapToLong(imd -> estimateSize(imd, context.info()))
+            .sum();
+
+        long storageSize = (long) (PERCENTAGE.get(configuration) * dataSetSize) / 100;
+        return new AutoscalingDeciderResult(AutoscalingCapacity.builder().total(storageSize, null).build(), new FrozenReason(dataSetSize));
+    }
+
+    static long estimateSize(IndexMetadata imd, ClusterInfo info) {
+        int copies = imd.getNumberOfReplicas() + 1;
+        return IntStream.range(0, imd.getNumberOfShards())
```

Comment:
🤨

Reply:
Are you requesting a change here? Will merge this as is but can change this if you wish in a follow-up.

Reply:
I changed it into a loop, guessing a bit that this was what you were after.